beacon_node/beacon_chain/src/light_client_optimistic_update_verification.rs

```diff
@@ -67,73 +43,36 @@ impl<T: BeaconChainTypes> VerifiedLightClientOptimisticUpdate<T> {
     pub fn verify(
-        light_client_optimistic_update: LightClientOptimisticUpdate<T::EthSpec>,
+        rcv_optimistic_update: LightClientOptimisticUpdate<T::EthSpec>,
         chain: &BeaconChain<T>,
         seen_timestamp: Duration,
     ) -> Result<Self, Error> {
-        let gossiped_optimistic_slot = light_client_optimistic_update.attested_header.beacon.slot;
-        let one_third_slot_duration = Duration::new(chain.spec.seconds_per_slot / 3, 0);
-        let signature_slot = light_client_optimistic_update.signature_slot;
-        let start_time = chain.slot_clock.start_of(signature_slot);
-        let mut latest_seen_optimistic_update = chain.latest_seen_optimistic_update.lock();
-
-        let head = chain.canonical_head.cached_head();
-        let head_block = &head.snapshot.beacon_block;
-        let attested_block_root = head_block.message().parent_root();
-        let attested_block = chain
-            .get_blinded_block(&attested_block_root)?
-            .ok_or(Error::FailedConstructingUpdate)?;
-
-        let attested_state = chain
-            .get_state(&attested_block.state_root(), Some(attested_block.slot()))?
-            .ok_or(Error::FailedConstructingUpdate)?;
-        let latest_seen_optimistic_update_slot = match latest_seen_optimistic_update.as_ref() {
-            Some(update) => update.attested_header.beacon.slot,
-            None => Slot::new(0),
-        };
-
-        // verify that no other optimistic_update with a lower or equal
-        // optimistic_header.slot was already forwarded on the network
-        if gossiped_optimistic_slot <= latest_seen_optimistic_update_slot {
-            return Err(Error::OptimisticUpdateAlreadySeen);
-        }
-
         // verify that enough time has passed for the block to have been propagated
-        match start_time {
-            Some(time) => {
-                if seen_timestamp + chain.spec.maximum_gossip_clock_disparity()
-                    < time + one_third_slot_duration
-                {
-                    return Err(Error::TooEarly);
-                }
-            }
-            None => return Err(Error::SigSlotStartIsNone),
-        }
-
-        // check if we can process the optimistic update immediately
-        // otherwise queue
-        let canonical_root = light_client_optimistic_update
-            .attested_header
-            .beacon
-            .canonical_root();
-
-        if canonical_root != head_block.message().parent_root() {
-            return Err(Error::UnknownBlockParentRoot(canonical_root));
+        let start_time = chain
+            .slot_clock
+            .start_of(rcv_optimistic_update.signature_slot)
+            .ok_or(Error::SigSlotStartIsNone)?;
+        let one_third_slot_duration = Duration::new(chain.spec.seconds_per_slot / 3, 0);
+        if seen_timestamp + chain.spec.maximum_gossip_clock_disparity()
+            < start_time + one_third_slot_duration
+        {
+            return Err(Error::TooEarly);
         }
```

Comment:
I had a brief discussion with @michaelsproul regarding the rationale behind re-processing the light client updates (added in https://github.com/sigp/lighthouse/pull/3799). 

The current logic is that if we receive the light client update before receiving the new block, this would result in the message being queued for re-processing later. The point is that we would still validate the gossip message later so it propagates, rather than telling gossip to IGNORE the message straight up, which might hinder their propagation.

With the new logic here, beacon node would compute the update after block import, and we would publish this on gossip, but the `IGNORE` validation result propagated earlier could still prevent this from being propgated?

It seems like we might want to keep this `UnknownBlockParentRoot` check and re-processing logic differentiate this with `InvalidLightClientOptimisticUpdate`.

Reply:
Note: the re-processing logic should then apply to both optimistic update and finality updates, both will be broadcast when signature slot advances https://github.com/ethereum/consensus-specs/blob/dev/specs/altair/light-client/full-node.md#create_light_client_optimistic_update

Reply:
Reverted dropping the reprocessing logic in https://github.com/sigp/lighthouse/pull/4969/commits/b41ed0759964990bec486e223cc3a25ae2c93377

However the re-processing logic should be changed. In unstable when the `BlockImported` event is received by the reprocessing queue the validation logic can then create the latest update and pass validation. However with this PR only when the `recompute_and_cache_light_client_updates` fn completes, the validation logic will pass.

Reply:
The optimistic message validation rules require an equality on the entire message, including the sync aggregate which is dependent on the block at the signature slot, not its parent. Since the validity of the message depends on our local view of the current head it's not useful to queue by branch (parent_root).

Consider a case with heavy forking where there are two valid proposers at slot N in their respective branches. If you receive two optimistic updates with the same signature slot and different attested headers you will only propagate the one that matches your local view. Similarly, a missed slot means there's no possible valid optimistic update.

So the reprocess queue for optimistic updates at slot N should wait to complete running `recompute_and_cache_light_client_updates` on _any_ branch for slot N and expire after 1 slot.

Reply:
Got it, so it looks like we need to rework most of the re-reprocessing logic for light client updates.

Perhaps we can do this:
1. Add a new `ReprocessQueueMessage` called `NewLightClientUpdate`
2. Send this event to reprocessing queue after we compute a new update (upon completion of `recompute_and_cache_light_client_updates`)
3. Move the existing event handling logic from [`BlockImported` match arm](https://github.com/sigp/lighthouse/blob/33976121601d9d31d26c1bb2e1f0676e9c4d24fc/beacon_node/beacon_processor/src/work_reprocessing_queue.rs#L691-L739) to a new match arm for  `InboundEvent::Msg(NewLightClientUpdate)`. This marked it as ready for re-processing
4. A re-processed LC update will then either get _ACCEPTED_ or _IGNORED_.

Any `QueuedLightClientUpdate` that is not reprocessed is expired after 12 seconds ([delay queue](https://github.com/sigp/lighthouse/blob/33976121601d9d31d26c1bb2e1f0676e9c4d24fc/beacon_node/beacon_processor/src/work_reprocessing_queue.rs#L611-L614))

Reply:
> Since the validity of the message depends on our local view of the current head it's not useful to queue by branch (parent_root).

Would it make more sense to queue it by `block_root`? and then anything that's not reprocessed within a slot is expired.

Reply:
I made an attempt to update the reprocessing logic based on the above: https://github.com/dapplion/lighthouse/pull/7

I've just came across @michaelsproul's [comment here](https://github.com/sigp/lighthouse/pull/3799#discussion_r1084699605) which might be relevant to the parent root check here. I don't fully understand it, but it's too late here now so I'll revisit tomorrow.

Reply:
Suggestion looks great, merged into this branch